be/src/connector/sink_memory_manager.cpp

```diff
@@ -0,0 +1,138 @@
+        if (victim && victim->first->get_written_bytes() > writer_and_stream.first->get_written_bytes()) {
+            continue;
+        }
+        partition = key;
+        victim = &writer_and_stream;
+    }
+    if (victim == nullptr) {
+        return false;
+    }
+
+    auto result = victim->first->commit();
+    _commit_func(result);
+    LOG(INFO) << "kill victim: " << victim->second->filename() << " size: " << result.file_statistics.file_size;
+    _candidates->erase(partition);
+    return true;
+}
+
+int64_t SinkOperatorMemoryManager::update_releasable_memory() {
+    int64_t releasable_memory = _io_poller->releasable_memory();
+    _releasable_memory.store(releasable_memory);
+    return releasable_memory;
+}
+
+int64_t SinkOperatorMemoryManager::update_writer_occupied_memory() {
+    int64_t writer_occupied_memory = 0;
+    for (auto& [_, writer_and_stream] : *_candidates) {
+        writer_occupied_memory += writer_and_stream.first->get_written_bytes();
+    }
+    _writer_occupied_memory.store(writer_occupied_memory);
+    return _writer_occupied_memory;
+}
+
+SinkMemoryManager::SinkMemoryManager(MemTracker* query_pool_tracker, MemTracker* query_tracker)
+        : _query_pool_tracker(query_pool_tracker), _query_tracker(query_tracker) {
+    _high_watermark_ratio = config::connector_sink_mem_high_watermark_ratio;
+    _low_watermark_ratio = config::connector_sink_mem_low_watermark_ratio;
+    _urgent_space_ratio = config::connector_sink_mem_urgent_space_ratio;
+}
+
+SinkOperatorMemoryManager* SinkMemoryManager::create_child_manager() {
+    _children.push_back(std::make_unique<SinkOperatorMemoryManager>());
+    auto* p = _children.back().get();
+    DCHECK(p != nullptr);
+    return p;
+}
+
+bool SinkMemoryManager::can_accept_more_input(SinkOperatorMemoryManager* child_manager) {
+    if (!_apply_on_mem_tracker(child_manager, _query_pool_tracker)) {
+        return false;
+    }
+    if (!_apply_on_mem_tracker(child_manager, _query_tracker)) {
+        return false;
+    }
+    return true;
+}
+
+int64_t SinkMemoryManager::_total_releasable_memory() {
+    int64_t total = 0;
+    std::for_each(_children.begin(), _children.end(), [&](auto& child) { total += child->releasable_memory(); });
+    return total;
+}
+
+int64_t SinkMemoryManager::_total_writer_occupied_memory() {
+    int64_t total = 0;
+    std::for_each(_children.begin(), _children.end(), [&](auto& child) { total += child->writer_occupied_memory(); });
+    return total;
+}
+
+bool SinkMemoryManager::_apply_on_mem_tracker(SinkOperatorMemoryManager* child_manager, MemTracker* mem_tracker) {
+    if (mem_tracker == nullptr || !mem_tracker->has_limit()) {
+        return true;
+    }
+
+    auto available_memory = [&]() { return mem_tracker->limit() - mem_tracker->consumption(); };
+    auto low_watermark = [&]() { return mem_tracker->limit() * _low_watermark_ratio; };
+    auto high_watermark = [&]() { return mem_tracker->limit() * _high_watermark_ratio; };
```

Comment:
for watermark we can only calc once

Reply:
Yes, but I think the overhead is negligible.